ingestion/src/main.rs

```diff
@@ -0,0 +1,222 @@
+use axum::{
+    extract::{FromRef, State},
+    response::Json,
+    routing::post,
+    Router,
+};
+use bb8::PooledConnection;
+use bb8_postgres::PostgresConnectionManager;
+use chrono::{DateTime, Utc};
+use futures::{stream::FuturesUnordered, StreamExt};
+use serde::Serialize;
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+use thiserror::Error;
+use tokio_postgres::NoTls;
+
+pub mod permissions;
+use permissions::{fetch_permits, ParamPermitTable, StationPermitTable};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("postgres returned an error: {0}")]
+    Database(#[from] tokio_postgres::Error),
+    #[error("database pool could not return a connection: {0}")]
+    Pool(#[from] bb8::RunError<tokio_postgres::Error>),
+    #[error("parse error: {0}")]
+    Parse(String),
+    #[error("RwLock was poisoned: {0}")]
+    Lock(String),
+    #[error("Could not read environment variable: {0}")]
+    Env(#[from] std::env::VarError),
+}
+
+type PgConnectionPool = bb8::Pool<PostgresConnectionManager<NoTls>>;
+
+pub type PooledPgConn<'a> = PooledConnection<'a, PostgresConnectionManager<NoTls>>;
+
+type ParamConversions = Arc<HashMap<String, (String, i32)>>;
+
+#[derive(Clone, Debug)]
+struct IngestorState {
+    db_pool: PgConnectionPool,
+    param_conversions: ParamConversions, // converts param codes to element ids
+    permit_tables: Arc<RwLock<(ParamPermitTable, StationPermitTable)>>,
+}
+
+impl FromRef<IngestorState> for PgConnectionPool {
+    fn from_ref(state: &IngestorState) -> PgConnectionPool {
+        state.db_pool.clone() // the pool is internally reference counted, so no Arc needed
+    }
+}
+
+impl FromRef<IngestorState> for ParamConversions {
+    fn from_ref(state: &IngestorState) -> ParamConversions {
+        state.param_conversions.clone()
+    }
+}
+
+impl FromRef<IngestorState> for Arc<RwLock<(ParamPermitTable, StationPermitTable)>> {
+    fn from_ref(state: &IngestorState) -> Arc<RwLock<(ParamPermitTable, StationPermitTable)>> {
+        state.permit_tables.clone()
+    }
+}
+
+pub struct Datum {
+    timeseries_id: i32,
+    timestamp: DateTime<Utc>,
+    value: f32,
+}
+pub type Data = Vec<Datum>;
+
+async fn insert_data(data: Data, conn: &mut PooledPgConn<'_>) -> Result<(), Error> {
+    let query = conn
+        .prepare(
+            "INSERT INTO public.data (timeseries, obstime, obsvalue) \
+                VALUES ($1, $2, $3) \
+                ON CONFLICT DO NOTHING", // TODO: figure out whether this should be nothing or update
```

Comment:
So, I talked to Søren since writing this, and should probably update the comment.

When obsinn wants to correct some data, it resends the message with the data changed. Hearing that you might think "ok, so let's `DO UPDATE` and be happy". HOWEVER, as is usual at met, there's a hidden abomination lying just beneath the surface. You see, the new message does not actually have to contain an entry for the thing it wants to correct, when you get a new message with a given [type_id, station_id, timestamp-range], it overrides all previous data for that [type_id, station_id, timestamp-range], and any params that previously had data for that, but aren't present in the new message should be deleted.

To fully comply with the behaviour obsinn expects, we would need to scan the database for any matching data every time obsinn sends a message, which really isn't great. This is probably much easier for kvalobs, since it organises itself around its observation_ids which I think map directly to obsinn messages, making it easier to find and delete what it needs to. FWIW ODA doesn't handle this, kvkafka just forwards everything to writer, which has `DO UPDATE` semantics.

This looks to me like a typical case of tight coupling between systems at met. This interface was designed specifically for obsinn->kvalobs, with little or no regard to other consumers. It really should have been foreseeable when obsinn was written that kvalobs was going to be replaced, and some concern should be given to making that easier. What makes it really so much worse is the lack of documentation of these quirks, It's one thing when an API has sharp edges which are well signposted so you can design for them in advance, but I could only find this out by asking Søren the right questions, and I only had the information needed to ask those questions after getting very into the weeds.

Long term we can either comply with the weirdness here, and take the performance penalty, or get obsinn to accept and use a new separate endpoint for deletions. The latter definitely seems like the better option, and Søren seemed pretty receptive to it. Short term though, I think `DO UPDATE` is good enough as long as we leave a comment explaining the situation so we can get back to it

Reply:
I've switched the behaviour to `DO UPDATE` and added a big comment explaining the situation